Origin: uniontech
Author: zhangfeilong@uniontech.com
Last-Update: 2021-04-15

--- a/Makefile.direct
+++ b/Makefile.direct
@@ -349,6 +349,7 @@
 cordtest$(EXEEXT): $(srcdir)/cord/tests/cordtest.c cords base_lib $(UTILS)
 	rm -f cordtest$(EXEEXT)
 	./if_mach SPARC DRSNX $(CC) $(CFLAGS) -o cordtest $(srcdir)/cord/tests/cordtest.c gc.a cord.a -lucb
+	./if_mach SW_64 LINUX $(CC) $(CFLAGS) -o cord/de $(srcdir)/cord/tests/de.c cord/cordbscs.o cord/cordxtra.o gc.a -lcurses `./threadlibs`
 	./if_mach HP_PA HPUX $(CC) $(CFLAGS) -o cordtest $(srcdir)/cord/tests/cordtest.c gc.a cord.a -ldld `./threadlibs`
 	./if_mach M68K AMIGA $(CC) $(CFLAGS) -UGC_AMIGA_MAKINGLIB -o cordtest $(srcdir)/cord/tests/cordtest.c gc.a cord.a `./threadlibs`
 	./if_not_there cordtest$(EXEEXT) || $(CC) $(CFLAGS) -o cordtest$(EXEEXT) $(srcdir)/cord/tests/cordtest.c gc.a cord.a `./threadlibs`
--- a/include/gc_tiny_fl.h
+++ b/include/gc_tiny_fl.h
@@ -49,7 +49,7 @@
 # if defined(__LP64__) || defined (_LP64) || defined(_WIN64) \
         || defined(__s390x__) \
         || (defined(__x86_64__) && !defined(__ILP32__)) \
-        || defined(__alpha__) || defined(__powerpc64__) \
+        || defined(__alpha__) || defined(__sw_64__) || defined(__powerpc64__) \
         || defined(__arch64__)
 #  define GC_GRANULE_BYTES 16
 #  define GC_GRANULE_WORDS 2
--- a/include/private/gcconfig.h
+++ b/include/private/gcconfig.h
@@ -419,6 +419,10 @@
 #   endif
 #   define mach_type_known
 # endif
+# if defined(__sw_64) || defined(__sw_64__)
+#   define SW_64
+#   define mach_type_known
+# endif
 # if defined(_AMIGA) && !defined(AMIGA)
 #   define AMIGA
 # endif
@@ -741,6 +745,8 @@
                     /*                 (SOLARIS, LINUX, DRSNX variants) */
                     /*             ALPHA      ==> DEC Alpha             */
                     /*                  (OSF1 and LINUX variants)       */
+                    /*             SW_64      ==> SUNWAY                */
+                    /*                  (LINUX variants)                */
                     /*             LOONGARCH  ==> Loongson LoongArch    */
                     /*                  (LINUX 32- and 64-bit variants) */
                     /*             M88K       ==> Motorola 88XX0        */
@@ -1908,6 +1914,27 @@
 #   endif
 # endif /* HP_PA */
 
+# ifdef SW_64
+#   define MACH_TYPE "SW_64"
+#   define ALIGNMENT 8
+#   define CPP_WORDSZ 64
+#   ifdef LINUX
+#       define OS_TYPE "LINUX"
+#       define LINUX_STACKBOTTOM
+#       ifdef __ELF__
+#         define SEARCH_FOR_DATA_START
+#         define DYNAMIC_LOADING
+#       else
+#           define DATASTART ((ptr_t)0x140000000)
+#       endif
+        extern int _end[];
+#       define DATAEND ((ptr_t)(_end))
+#       define MPROTECT_VDB
+                /* Has only been superficially tested.  May not */
+                /* work on all versions.                        */
+#   endif
+# endif
+
 # ifdef ALPHA
 #   define MACH_TYPE "ALPHA"
 #   define ALIGNMENT 8
